### PR TITLE
ipaddress.IPAddress support

### PR DESCRIPTION
**Pull request by pilosus**
## Change Summary

``ipaddress``-related types added.
I'm not quite sure if my code really does exactaly what's needed in #333 or not. Please review the PR. I'm happy to make any changes to make things work as expected.
If the PR is okay, I'm willing to implement ``ipaddress.ip_network`` and ``ipaddress.ip_interface`` too.

## Related issue number

#333 

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

**Comment by samuelcolvin** [comment]
please fix tests and I'll review, see [here](https://pydantic-docs.helpmanual.io/#contributing-to-pydantic) for details.

File: pydantic/types.py [hunk 1]
```diff
@@ -28,39 +29,42 @@
...
-    'condecimal',
-    'UUID1',
-    'UUID3',
-    'UUID4',
-    'UUID5',
-    'FilePath',
-    'DirectoryPath',
-    'Json',
-    'JsonWrapper',
+    "NoneStr",
```
Comment (samuelcolvin) [inline pydantic/types.py]: please revert all these quotes changes and use single quotes in all code.

Comment (pilosus) [reply pydantic/types.py]: Sorry I had used black formatter the wrong way. 
I've just reverted these changes

File: pydantic/types.py [hunk 2]
```diff
@@ -497,3 +503,46 @@ def validate(cls, v: str) -> Any:
...
+    def validate(cls, value: IPAddr) -> str:
+        try:
+            ip = cls._constructor(value)
+        except ValueError:
+            raise cls._exc()
+        else:
+            return str(ip)
+
+
+class IPv4Address(IPAddress):
```
Comment (samuelcolvin) [inline pydantic/types.py]: Hi, I'm afraid this isn't the right way of implementing IP address verification.

The point is that the stdlib version of [`IPv4Address`](https://docs.python.org/3/library/ipaddress.html#ipaddress.IPv4Address) and `IPv6Address` should work as type annotations.

The only custom class we should implement is `IPvAnyAddress` which can be either a v4 or v6 address, as per #333 this should inherit from `_BaseAddress`, but `validate` should return either a `IPv4Address` or `IPv6Address`.

Have a look at [this code](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/validators.py#L267-L287) for how we validate standard lib types.

Comment (pilosus) [reply pydantic/types.py]: Okay, now I see what's should be done.

But shall the ``validate`` take in either ``str``, ``bytes`` or ``int`` just like ``ipaddress.ip_address`` does?

Comment (samuelcolvin) [reply pydantic/types.py]: yes, argument annotation should be `Union[str, bytes, int]`

Comment (pilosus) [reply pydantic/types.py]: I've just submitted a new implementation.

File: pydantic/types.py [hunk 3]
```diff
@@ -497,3 +500,23 @@ def validate(cls, v: str) -> Any:
...
+
+    @classmethod
+    def validate(cls, value: Union[str, bytes, int]) -> Union[IPv4Address, IPv6Address]:
+        try:
+            return IPv4Address(value)
+        except (AddressValueError, NetmaskValueError):
+            pass
+
+        try:
+            return IPv6Address(value)
```
Comment (samuelcolvin) [inline pydantic/types.py]:
```py
with change_exception(errors.IPvAnyAddressError, ValueError):
    return IPv6Address(value)
```

File: pydantic/errors.py [hunk 4]
```diff
@@ -288,3 +288,15 @@ class DataclassTypeError(PydanticTypeError):
 
 class CallableError(PydanticTypeError):
     msg_template = '{value} is not callable'
+
+
+class IPvAnyAddressError(PydanticTypeError):
```
Comment (samuelcolvin) [inline pydantic/errors.py]:
```suggestion
class IPvAnyAddressError(PydanticValueError):
```

File: pydantic/errors.py [hunk 5]
```diff
@@ -288,3 +288,15 @@ class DataclassTypeError(PydanticTypeError):
...
+    msg_template = 'value is not a valid IPv4 or IPv6 address'
+
+
+class IPv4AddressError(PydanticTypeError):
```
Comment (samuelcolvin) [inline pydantic/errors.py]:
```suggestion
class IPv4AddressError(PydanticValueError):
```

File: pydantic/errors.py [hunk 6]
```diff
@@ -288,3 +288,15 @@ class DataclassTypeError(PydanticTypeError):
...
+    msg_template = 'value is not a valid IPv4 address'
+
+
+class IPv6AddressError(PydanticTypeError):
```
Comment (samuelcolvin) [inline pydantic/errors.py]:
```suggestion
class IPv6AddressError(PydanticValueError):
```

File: pydantic/types.py [hunk 7]
```diff
@@ -497,3 +500,23 @@ def validate(cls, v: str) -> Any:
...
+class IPvAnyAddress(_BaseAddress):
+    @classmethod
+    def __get_validators__(cls) -> 'CallableGenerator':
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value: Union[str, bytes, int]) -> Union[IPv4Address, IPv6Address]:
+        try:
+            return IPv4Address(value)
+        except (AddressValueError, NetmaskValueError):
```
Comment (samuelcolvin) [inline pydantic/types.py]:
```suggestion
        except ValueError:
```

Or change everywhere so we're consistent.

File: pydantic/types.py (hunk 3 above)
Comment (pilosus) [reply pydantic/types.py]: Great suggestions, thanks! I've fixed them all. Hope now all goes well

File: pydantic/validators.py [hunk 8]
```diff
@@ -218,6 +219,22 @@ def decimal_validator(v: Any) -> Decimal:
     return v
 
 
+def ip_v4_address_validator(v: Any) -> IPv4Address:
+    if isinstance(v, IPv4Address):
+        return v
```
Comment (samuelcolvin) [inline pydantic/validators.py]: no test for this case

File: pydantic/validators.py [hunk 9]
```diff
@@ -218,6 +219,22 @@ def decimal_validator(v: Any) -> Decimal:
...
+
+    with change_exception(errors.IPv4AddressError, ValueError):
+        return IPv4Address(v)
+
+
+def ip_v6_address_validator(v: Any) -> IPv6Address:
+    if isinstance(v, IPv6Address):
+        return v
```
Comment (samuelcolvin) [inline pydantic/validators.py]: no test for this case

File: pydantic/validators.py (hunk 8 above)
Comment (pilosus) [reply pydantic/validators.py]: Oh, you're right! I've fixed both cases. Now coverage's back to almost 100%

**Comment by samuelcolvin** [comment]
Awesome. Thank you very much.

**Pull request merged**